File: examples/python/patterns/determinism/non-deterministic-in-step.py
import uuid
from aws_durable_execution_sdk_python import DurableContext, durable_execution, durable_step
from aws_durable_execution_sdk_python.types import StepContext


@durable_step
def generate_transaction_id(ctx: StepContext) -> str:
    return str(uuid.uuid4())


@durable_step
def charge(ctx: StepContext, amount: float, transaction_id: str) -> dict:
    return payment_service.charge(amount, transaction_id)


@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    transaction_id = context.step(generate_transaction_id())
    receipt = context.step(charge(event["amount"], transaction_id))
    return {"transactionId": transaction_id, "receipt": receipt}

File: examples/csharp/patterns/determinism/non-deterministic-in-step.cs
using Amazon.Lambda.Core;
using Amazon.Lambda.DurableExecution;

public class NonDeterministicInStepExample
{
    public Task<DurableExecutionInvocationOutput> Handler(
        DurableExecutionInvocationInput input, ILambdaContext context)
        => DurableFunction.WrapAsync<ChargeInput, ChargeResult>(Workflow, input, context);

    private async Task<ChargeResult> Workflow(ChargeInput input, IDurableContext ctx)
    {
        // Wrap the non-deterministic call in a step so the value is checkpointed.
        string transactionId = await ctx.StepAsync(
            async (_, _) => Guid.NewGuid().ToString(),
            name: "generate-transaction-id");

        Receipt receipt = await ctx.StepAsync(
            async (_, _) => Charge(input.Amount, transactionId),
            name: "charge");

        return new ChargeResult(transactionId, receipt);
    }

    private static Receipt Charge(decimal amount, string transactionId)
        => new(transactionId, amount);
}

public record ChargeInput(decimal Amount);
public record Receipt(string TransactionId, decimal Amount);
public record ChargeResult(string TransactionId, Receipt Receipt);
